### PR TITLE
Dsp3/zephyr 3.2

### PR DESCRIPTION
**Pull request by lbonhomme-vossloh**
- The option to configure the write flash alignment  in mcuboot.
- All I2S fixes to work on STM32H743ZI
- The full-duplex integration and its Zephyr's config option.
- The resistances of TX under-run and RX overrun and their zephyr's config option.
- The IO swap option for I2S configuration

### CODE REVIEW

File: drivers/i2s/Kconfig.stm32 [hunk 1]
```diff
@@ -69,4 +69,11 @@ config I2S_STM32_FULL_DUPLEX
 	  Full-duplex mode is compatible on I2S
 	  The board support the I2S in full-duplex mode
 
+config I2S_STM32_RX_OVERRUN_DROP
+	bool "Drop data received in overrun context"
+	default n
+	help
+	  Drop data received in overrun context
```
Comment (smartin-vossloh) [inline drivers/i2s/Kconfig.stm32:76]: @lbonhomme-vossloh  nit: missing dot at the end.

ditto below

File: drivers/i2s/i2s_ll_stm32.c [hunk 2]
```diff
@@ -788,18 +816,46 @@ static int rx_stream_start_dma(struct stream *stream, const struct device *dev,
 static int tx_stream_start_dma(struct stream *stream, const struct device *dev, bool full_duplex)
 {
 	const struct i2s_stm32_cfg *cfg = dev->config;
-	size_t mem_block_size;
 	int ret;
 
+#ifdef CONFIG_I2S_STM32_TX_UNDERRUN_LAST_REPEAT
+	/* Initiate  the block to send in underrun context*/
```
Comment (smartin-vossloh) [inline drivers/i2s/i2s_ll_stm32.c:833]: @lbonhomme-vossloh nit: `s/Initiate  the/Initialize the/`

File: drivers/i2s/i2s_ll_stm32.c [hunk 3]
```diff
@@ -254,6 +254,11 @@ static int i2s_stm32_configure(const struct device *dev, enum i2s_dir dir,
 		LL_I2S_DisableMasterClock(cfg->i2s);
 	}
 
+	/* set I2S */
+	if(i2s_cfg->options & I2S_OPT_IO_SWAP) {
```
Comment (smartin-vossloh) [inline drivers/i2s/i2s_ll_stm32.c:258]: @lbonhomme-vossloh missing space after keyword before `(`

File: drivers/i2s/i2s_ll_stm32.c [hunk 4]
```diff
@@ -345,25 +345,28 @@ static int i2s_stm32_read(const struct device *dev, void **mem_block,
...
-		if (ret < 0) {
-			return ret;
-		}
+	if (dev_data->rx.state == I2S_STATE_ERROR) {
+		LOG_ERR("Error status");
+		return -EIO;
+	}
+
+	ret = k_sem_take(&dev_data->rx.sem,
+				SYS_TIMEOUT_MS(dev_data->rx.cfg.timeout));
```
Comment (smartin-vossloh) [inline drivers/i2s/i2s_ll_stm32.c:358]: @lbonhomme-vossloh mis-alignment

File: drivers/i2s/i2s_ll_stm32.c [hunk 5]
```diff
@@ -874,6 +886,11 @@ static void tx_stream_disable(struct stream *stream, const struct device *dev)
 {
 	const struct i2s_stm32_cfg *cfg = dev->config;
 
+#ifdef CONFIG_SOC_SERIES_STM32H7X
+	LL_SPI_SuspendMasterTransfer(cfg->i2s);
+#endif
+
+
```
Comment (smartin-vossloh) [inline drivers/i2s/i2s_ll_stm32.c]: @lbonhomme-vossloh 1 blank line is enough ;)

File: drivers/i2s/i2s_ll_stm32.c [hunk 6]
```diff
@@ -510,14 +515,32 @@ static void dma_rx_callback(const struct device *dma_dev, void *arg,
...
 		goto rx_disable;
 	}
-	k_sem_give(&stream->sem);
+
+#ifdef CONFIG_I2S_STM32_RX_OVERRUN_DROP
+	if(rx_overrun == false) {
+		k_sem_give(&stream->sem);
+	}
+#else
+		k_sem_give(&stream->sem);
```
Comment (smartin-vossloh) [inline drivers/i2s/i2s_ll_stm32.c:548]: @lbonhomme-vossloh weird indentation

File: drivers/i2s/i2s_ll_stm32.c [hunk 7]
```diff
@@ -984,6 +1040,10 @@ static void tx_stream_disable(struct stream *stream, const struct device *dev)
 		stream->mem_block = NULL;
 	}
 
+#ifdef CONFIG_I2S_STM32_TX_UNDERRUN_LAST_REPEAT
+		k_mem_slab_free(stream->cfg.mem_slab, &stream->last_mem_block);
```
Comment (smartin-vossloh) [inline drivers/i2s/i2s_ll_stm32.c:1055]: @lbonhomme-vossloh weird indentation

ditto below (twice)

**Review by nvincent-vossloh — CHANGES_REQUESTED** [review]

File: drivers/i2s/Kconfig.stm32 [hunk 8]
```diff
@@ -62,4 +62,25 @@ config I2S_STM32_PLLI2S_PLLR
 	  frequency is less than or equal to 192MHz.
 	  Allowed values: 2-7
 
+config I2S_STM32_FULL_DUPLEX
+	bool "Full-duplex mode is compatible on I2S"
+	default n
+	help
+	  Full-duplex mode is compatible on I2S.
+	  The board support the I2S in full-duplex mode.
```
[the remark this answers is not in the document]
Comment (lbonhomme-vossloh) [reply drivers/i2s/Kconfig.stm32:70]: This flag doesn't set the full duplex mode on I2S. The flag authorizes only to set I2S in full duplex mode.
The direction mode (RX, TX or both) is set by zephyr API and I2S_config structure.

**Review by nvincent-vossloh — APPROVED** [review]
ok, wait @smartin-vossloh approval before merging.

**Review by nvincent-vossloh — APPROVED** [review]
Looking good.

**Review by smartin-vossloh — APPROVED** [review]
lgtm

**Pull request merged**